src/librustc_typeck/check/pat.rs

```diff
@@ -1409,18 +1424,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 if min_len != len {
                     self.error_scrutinee_inconsistent_length(span, min_len, len);
                 }
-            } else if let r @ Some(_) = len.checked_sub(min_len) {
+            } else if let Some(pat_len) = len.checked_sub(min_len) {
                 // The variable-length pattern was there,
                 // so it has an array type with the remaining elements left as its size...
-                return r;
+                return Some(PatLenOrArray::Pat(pat_len));
             } else {
                 // ...however, in this case, there were no remaining elements.
                 // That is, the slice pattern requires more than the array type offers.
                 self.error_scrutinee_with_rest_inconsistent_length(span, min_len, len);
             }
+        } else if slice.is_none() {
+            // We have a pattern with a fixed length,
+            // which we can use to infer the length of the array.
+            // of the array.
+            let updated_arr = self.tcx.mk_array(inner_ty, min_len);
+            self.demand_eqtype(span, updated_arr, arr_ty);
+            return Some(PatLenOrArray::Array(updated_arr));
```

Comment:
Is this extra complexity (compared to just unifying the constant) in order to be able to use `demand_eqtype`?

Reply:
tbh I wasn't able to find a good way to unify consts.
The only thing I found was [`unify_const_variable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.unify_const_variable), which requires a `ConstVid`, which I do not fully understand yet. At this point I gave up and just used type equality :laughing: 

So while this is the least complex thing I found there could easily be a better approach.

Reply:
The current approach might do some extra work, but it seems fairly clean, and it can't go wrong. We could always improve this later imo.

Reply:
@lcnr That's mostly because there aren't specific functions, it's all done through traits.

`demand_eqtype`, for example, passes `actual` and `expected` to a function that can take just about anything.